hc_core_api/src/lib.rs

```diff
@@ -0,0 +1,263 @@
+/*!
+hc_core_api provides a library for container applications to instantiate and run holochain applications.
+
+# Examples
+
+``` rust
+extern crate hc_core;
+extern crate hc_core_api;
+extern crate hc_dna;
+extern crate hc_agent;
+
+use hc_core_api::*;
+use hc_dna::Dna;
+use hc_agent::Agent;
+use std::sync::{Arc, Mutex};
+use hc_core::context::Context;
+use hc_core::logger::SimpleLogger;
+use hc_core::persister::SimplePersister;
+
+// instantiate a new app
+
+// need to get to something like this:
+//let dna = hc_dna::from_package_file("mydna.hcpkg");
+
+// but for now:
+let dna = Dna::new();
+let agent = Agent::from_string("bob");
+let context = Context {
+    agent: agent,
+    logger: Arc::new(Mutex::new(SimpleLogger {})),
+    persister: Arc::new(Mutex::new(SimplePersister {})),
+};
+let mut hc = Holochain::new(dna,Arc::new(context)).unwrap();
+
+// start up the app
+hc.start().expect("couldn't start the app");
+
+// call a function in the app
+hc.call("some_fn");
+
+// get the state
+{
+    let state = hc.state();
+
+    // do some other stuff with the state here
+    // ...
+}
+
+// stop the app
+hc.stop().expect("couldn't stop the app");
+
+```
+*/
+
+extern crate hc_agent;
+extern crate hc_core;
+extern crate hc_dna;
+
+use hc_core::context::Context;
+use hc_dna::Dna;
+use std::sync::Arc;
+
+/// contains a Holochain application instance
+#[derive(Clone)]
+pub struct Holochain {
+    instance: hc_core::instance::Instance,
+    context: Arc<hc_core::context::Context>,
+    active: bool,
+}
+
+use hc_core::error::HolochainError;
+use hc_core::nucleus::fncall;
+use hc_core::nucleus::Action::*;
+use hc_core::state::Action::*;
+use hc_core::state::State;
+
+impl Holochain {
+    /// create a new Holochain instance
+    pub fn new(dna: Dna, context: Arc<Context>) -> Result<Self, HolochainError> {
+        let mut instance = hc_core::instance::Instance::new();
+        let action = Nucleus(InitApplication(dna.clone()));
+        instance.dispatch(action);
+        instance.consume_next_action()?;
+        context.log("instantiated")?;
+        let app = Holochain {
+            instance: instance,
+            context: context,
+            active: false,
+        };
+        Ok(app)
+    }
+
+    /// activate the Holochain instance
+    pub fn start(&mut self) -> Result<(), HolochainError> {
+        if self.active {
+            return Err(HolochainError::InstanceActive);
+        }
+        self.active = true;
+        Ok(())
+    }
+
+    /// deactivate the Holochain instance
+    pub fn stop(&mut self) -> Result<(), HolochainError> {
+        if !self.active {
+            return Err(HolochainError::InstanceNotActive);
+        }
+        self.active = false;
+        Ok(())
+    }
+
+    /// call a function in a zome
+    pub fn call(&mut self, fn_name: &str) -> Result<(), HolochainError> {
```

Comment:
What about function parameters?

Reply:
yep, just didn't get to it...

Reply:
also return value needs to be the encapsulated return value from the function call...

Reply:
As I said verbally, I am implementing all the stuff you need so that we can actually call zome functions. For that I have to change the API offered by Instance. So the body of this function will have to change anyway. But I would like to get your changes in first because of the renaming you did. So I would suggest we merge this branch first without having this working actually and then I go and merge this (from develop) into my branch and create a PR for that..

So, what I say is: make the signature of this final but don't bother implementing it in a functional way.